File: beginner/greedy_method/saruman_army/c++/sarumanArmy.cpp
#include <algorithm>

using namespace std;
// ===== ↓↓↓input↓↓↓ ===== //
int N = 6;
int R = 10;
int X[6] = {1, 7, 15, 20, 30, 50}; 
// ===== ↑↑↑input↑↑↑ ===== //

int main(int argc, char *argv[]) {
	sort(X, X + N);

    int i = 0, ans = 0;
    while (i < N) {
        // s is not cover leftest point
        int s = X[i++];

        while (i < N && X[i] <= s + R) i++;

        // p is next point
        int p = X[i - 1];
        
        // move until p is beyond R
        while (i <  N && X[i] <= p + R) i++;

        ans++;
    }
    printf("%d\n", ans);
}

File: beginner/greedy_method/saruman_army/go/sarumanArmy.go
package main

import (
	"fmt"
	"sort"
)

// ===== ↓↓↓input↓↓↓ ===== //

var N int = 6
var R int = 10
var X = []int{1, 7, 15, 20, 30, 50}

// ===== ↑↑↑input↑↑↑ ===== //

func main() {
	sort.Ints(X)

	i := 0
	ans := 0
	for i < N {
		i++
		s := X[i]

		for i < N && X[i] <= s+R {
			i++
		}

		p := X[i-1]

		for i < N && X[i] <= p+R {
			i++
		}

		ans++
	}
	fmt.Print(ans)
}
